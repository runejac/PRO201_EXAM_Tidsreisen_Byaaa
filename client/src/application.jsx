import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
