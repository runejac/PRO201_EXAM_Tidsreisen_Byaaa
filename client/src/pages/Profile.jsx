import React, { useContext } from "react";
import { NotLoggedIn } from "../components/NotLoggedIn.jsx";
import "../css/profile.css";
import { IoPersonOutline } from "react-icons/io5";
import { User } from "../application.jsx";
import { useNavigate } from "react-router-dom";
import { CapsuleButtonLetthet } from "../components/CapsuleButton.jsx";
import { fetchJSON_client } from "../helpers/http.jsx";
import { FiLogOut } from "react-icons/fi";

function ProgressBar({ color, progress }) {
  const fillerStyles = {
    height: "100%",
    maxWidth: "15rem",
    width: `${progress / 14}rem`,
    backgroundColor: color,
    borderRadius: "inherit",
    textAlign: "right",
  };

  return (
    <div id="progressBar">
      <div style={fillerStyles}></div>
    </div>
  );
}

export function Profile() {
  const { user, setUser } = useContext(User);
  const { name, points, finishedCapsules } = user;
  const navigate = useNavigate();

  const logOutHandler = async () => {
    if (confirm("Er du sikker på at du ønsker å logge ut?")) {
      navigate("/");
      await fetchJSON_client("/api/deleteCookie");
      return true;
    } else {
      return false;
    }
  };

  if (name === undefined) {
    return <NotLoggedIn />;
  }

  return (
    <div className="profileContainer">
      <div id="main">
        <div id={"container"}>
          <div id={"logout-container"}>
            <div id={"logout-link"} onClick={logOutHandler}>
              <FiLogOut /> <span>Logg ut</span>
            </div>
          </div>
          <div id={"user-information-container"}>
            <h1 id="userName">{name}</h1>
            <div id="circle">
              <IoPersonOutline id="icon" />
            </div>
            <ProgressBar color="#333333" progress={points} />
            <p id="userPoints">{points}/210 poeng</p>
            <div id={"myfindings-button-container"}>
              <CapsuleButtonLetthet
                onClick={() => navigate("/myfindings")}
                buttonText={"Mine funn"}
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
