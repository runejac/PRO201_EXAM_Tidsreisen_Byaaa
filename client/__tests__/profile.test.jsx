import React from "react";
import { MemoryRouter } from "react-router-dom";
import { Profile } from "../src/pages/Profile.jsx";
import renderer from "react-test-renderer";
import { User } from "../src/application.jsx";

describe("Profile", () => {
  it("shows snapshot", async () => {
    const user = { name: "Hei", points: 23, finishedCapsules: [] };
    const component = renderer.create(
      <MemoryRouter>
        <User.Provider value={{ user }}>
          <Profile />
        </User.Provider>
      </MemoryRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
