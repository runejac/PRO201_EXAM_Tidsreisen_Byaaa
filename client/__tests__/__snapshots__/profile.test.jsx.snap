// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile shows snapshot 1`] = `
<div
  className="profileContainer"
>
  <div
    id="main"
  >
    <div
      id="container"
    >
      <div
        id="logout-container"
      >
        <div
          id="logout-link"
          onClick={[Function]}
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
            />
            <polyline
              points="16 17 21 12 16 7"
            />
            <line
              x1="21"
              x2="9"
              y1="12"
              y2="12"
            />
          </svg>
           
          <span>
            Logg ut
          </span>
        </div>
      </div>
      <div
        id="user-information-container"
      >
        <h1
          id="userName"
        >
          Hei
        </h1>
        <div
          id="circle"
        >
          <svg
            fill="currentColor"
            height="1em"
            id="icon"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M344 144c-3.92 52.87-44 96-88 96s-84.15-43.12-88-96c-4-55 35-96 88-96s92 42 88 96z"
              fill="none"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="32"
            />
            <path
              d="M256 304c-87 0-175.3 48-191.64 138.6C62.39 453.52 68.57 464 80 464h352c11.44 0 17.62-10.48 15.65-21.4C431.3 352 343 304 256 304z"
              fill="none"
              strokeMiterlimit="10"
              strokeWidth="32"
            />
          </svg>
        </div>
        <div
          id="progressBar"
        >
          <div
            style={
              Object {
                "backgroundColor": "#333333",
                "borderRadius": "inherit",
                "height": "100%",
                "maxWidth": "15rem",
                "textAlign": "right",
                "width": "1.6428571428571428rem",
              }
            }
          />
        </div>
        <p
          id="userPoints"
        >
          23
          /210 poeng
        </p>
        <div
          id="myfindings-button-container"
        >
          <button
            id="button-capsules-letthet"
            onClick={[Function]}
          >
            Mine funn
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
